[WIP] Infection Gamemode (upgraded blob + war ops styled)

Whoneedspacee opened the pull request:
## About The Pull Request

**Credits to @FuryMcFlurry for the gamemodes icons, wouldn't be nearly as cool without him**

Adds a new antagonist, The Infection. When spawned, the station is alerted, and a certain amount of time later (most likely 10 minutes) the core lands a bit off the station. 

There are also beacon generators which generate barriers that prevent the infection from crossing them and will instantly destroy any that do manage to cross it, only taking damage from the infection expanding onto it.

The new infection has many different things, like starting off with at least one evolving spore controlled by a player. These are quite like blobbernauts in blob but they respawn after a certain amount of time after they die, and can select any node to respawn from.

They also are allowed to upgrade and change into different paths in order to counter certain station tactics and avoid situations like emitter spam being the be all end all for winning the game. The infection itself also possesses things like this in the form of being able to upgrade many of their structures, and spend upgrade points gained from destroying the beacons defending the station to gain special abilities or create more evolving spore allies.

A direct reason for this gamemodes existence is preventing early and lame endings to station combats such as nuclear operatives, or situations where the objective can be easily rushed and the round can be over in a short time without many noticing. Which is why the core is immune to all forms of damage or destruction other than from legendary weapons.

Legendary weapons are garnered from killing bosses in events which spawn every 10 minutes starting when the core of the infection lands. They're generally special weapons with special abilities that make it easier to fight the infection or provide support for allies. 

A big thing to note as well is that unlike blob, humans are fully able to move through the infection (with a speed reduction)

## Why It's Good For The Game

This game is generally lacking in a gamemode with a teamwork oriented objective that cannot be rushed. It provides a unique outlet in which you are well aware that everyone around you on the station is your ally, and can work together to your full potential.

There are many things that need to be achieved for a successful round and most will typically involve requiring help of others. If you want to have science past the 10 minute mark you will most likely need to move the machines further back into the station.

If you want power, engineering will likely have to set up generators in order to have power once engineering is taken over. Overall I can't predict how successful a gamemode will be, how fun it will be, or the effort that people are willing to put into it, but I've never seen a gamemode with this much team oriented effort yet, and I think it's a unique opportunity for people to build mechs and create tons of botany guns or other chemicals for people without it being for the goal of destroying something.

Pretty much everyone can try as hard as they want to powergame as much as they can to fight the infection since it has the unique factor of being a clear objective unlike most gamemodes in this game.

**Current Version Video Infection v7**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
https://www.youtube.com/watch?v=s1a1o8RxLpw&feature=youtu.be
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**

## Changelog
:cl: Joe Berry
add: Added The Infection as an antagonist and a gamemode. This is a station defense gamemode against a creature like the blob, except much stronger with upgrades and player controlled minions.
config: Added infection to the rotation with a weight equal to nuclear operatives.
/:cl:

Whoneedspacee committed:
ports over all the old infection stuff to this new branch

adds infection, basically a blob rip off that will have different upgrade paths

adds beacons, block infection advancement, can only be destroyed by the infection structures directly trying to consume it

adds legendary weapons, the only things that can destroy the infection core, will be dropped by bosses

icons and sounds and whatnot for everything

adds landmarks for beacon spawns and infection spawns, beacons have directions they face as well, they defend everything behind them with a wall that blocks infection

Whoneedspacee committed:
moves beacons around to promote more teamwork positions, out in the open where everyone can guard them.

beacons now spawn when the commander mob spawns in, not when the core lands.

improves cinematic feel, the biohazard containment commander now talks to you, gives cues for events that are happening, and tells you when beacons have been destroyed, as well as gives hints about how to kill the infection.

blobber- i mean infesternauts no longer rely on factories to live anymore, and just need to stay near the infection.

infesternauts are now deleted on death, will add respawning to them later.

resource infections now only add 1 point per life proc.

the infection can now evolve, get permanent improvements, 6 at the start of the game and 1 more per beacon destroyed.

Whoneedspacee committed:
removes manual expansion, clicking now tries to open a menu on structures to do certain things to them

adds base for infection structures to be upgraded

fixes up flavortext to be less wall of texty

everything is radial menu's now because they're cool

tones down infection node default expansion rate

Whoneedspacee committed:
adds radial icons for infection menus

moves upgrading stuff to the base infection file

adds upgrade types and extra descriptions for what the upgrades do for infection structures

adds base of upgrading stuff for infection mobs and removes infesternauts

adds maximum upgrade levels for evolution

fixes flavortext and makes stuff timed properly

spores get upgrade points beacons are destroyed and start off with the maximum gotten so far

Whoneedspacee committed:
why was this not versioned

Whoneedspacee committed:
naming changes to remove blob words

no more noises when beacons get hit for the commander

increases max starting points for the infection

adds upgraded shields, max upgraded shields spawn around the core now

reduces factory upgrade cost because factories suck

reduces node upgrade cost and gives them some more buffs to make expansion less of a hassle

improves resource upgrade generation

shields upgrade into reflective blobs and then stronger reflective blobs

updates help text for infection commander

changes infection light color to be cooler looking

normal infection no longer spawn upgrade menus

icon stuff added

Whoneedspacee committed:
adds slowdown when moving in infection for non blob creatures

adds turret infection, automatically fire at nearby enemies to the infection

increases base point rate for core

doubles core expansion rate

the core now automatically builds max level turrets

spores now properly recieve their upgrade points

non factory spores (sentient spores) can no longer zombify bodies

adds new icons for turret infection

Whoneedspacee committed:
overmind check for turrets so they don't fire repeatedly even when the overmind is dead

atmos pass stuff for infection structures

Whoneedspacee committed:
increases default commander upgrade points

the core now shakes the screen when it is attacked with an effective attack

the descriptions for the special infections are now more descriptive for what they are doing and what the upgrade will do

cuts down on worthless code for the evolve menu and adds a new future section for passive effects / misc stuff

the infection now takes more explosive damage dependent on the severity of the explosion

adds a new icon for the radial menus, a plus sign used for structural upgrades

Whoneedspacee committed:
reverses orientation of beacons and infection spawn on map

fixes upgrade text for everything

fixes lighting issues

infection no longer expands to space

Whoneedspacee committed:
moves infection icons to blob icons because we werent really using it anyways and they already share a bunch of shit

reworks infection upgrading alltogether to be custom per structure

fixes infection layers to not overlay a bunch of shit it shouldnt

adds ambience to infection on entry

the infection no longer absolutely destroys mechs

adds 3 turret types, resistant, infernal, and homing.

adds framework for shields to be upgradeable through non-radial-menu ways (so it doesnt take 3000 years)

removes levels and max levels for infection structures because it was stupid and boring

infection turret bullets can no longer go past beacon walls

Whoneedspacee committed:
sets up datum base for a wayyyyyyyy easier to manage upgrades system

adds can_upgrade proc as the upgrade point version of can_buy

Whoneedspacee committed:
removes show_description in all the upper level files, probably to be replaced by a standardized version in the default file.

turret upgrades now work, homing upgrades are mostly finished, might be changed a little bit in the future.

renames upgrades list to upgrade list

adds flak bullets icon

Whoneedspacee committed:
core turrets are now max upgraded resistance turrets with faster speed bullets

infection mobs now target the dead and also now zombify properly if spawned without a factory

upgrade procs for the turrets now recieve the parent structure

adds upgrades for every turret type

client spawned spores no longer zombify

upgrades now have a do_upgrade proc instead of being snowflake handled in the upgrade menu

Whoneedspacee committed:
sets up the base for sentient spore types and upgrades

adds ambiguity to upgrades datum so it can be used for multiple things

the infection can no longer use the evolve menu before it is placed (woo no bugs)

fixes text not displaying when attempting to upgrade a already max upgraded structure

adds really bad icons for the mob types

Whoneedspacee committed:
adds sentient subtype for infection spores (why didnt i do this originally this is the problem with coding at 2 am)

adds respawn points and respawn timers for sentient infection spores

tries to fix type checking for similar infection types nearby (wouldn't stop defensive turrets from being placed next to normal turrets, which could be upgraded into defensive turrets, etc...)

show_description now does something for infection mobs and structures (displays upgrade names and their descriptions)

create_spore is now called async so it doesn't cancel other calls

Whoneedspacee committed:
adds an ishelper for infection mobs

fixes communication between infection mobs

removes useless respawn point type and replaces it with the core itself

infection mobs now have a stat for their upgrade points

infection mob respawn points can now be any node and if there are no nodes, any infection alive

infection mobs can now smash walls if they're the right type

more descriptive upgrades and adds a non-functional speed boost upgrades

create_spore no longer steals one of the commanders points

Whoneedspacee committed:
spores may now refund their upgrades

spores may now upgrade while respawning and may only upgrade next to the core or nodes

adds a new proc for spores to change their type

adds upgrade complete messages

the infection now takes more damage from bombs (it really was pitiful)

Whoneedspacee committed:
adds infection bosses file for future setup

respawning spores can now alternate their set location

changes core defense setup

adds smokey particles to the core

adds armor penetration to core turrets

removes speed upgrade since new types will override the old stuff

knockback turrets no longer throw back mechs and stronger moving atoms

fixes infection structure type checking

spores no longer spawn with the incorrect amount of upgrade points

removes ambience since it didnt even work properly (stacked with area ambience) and will be replaced with just general sounds for ambience or i'll figure out how to change the area

Whoneedspacee committed:
last update for a while maybe

Whoneedspacee committed:
adds comsigs for shit i need

upgrades are now handled by component datums basically instead of just datums because im a retard and now the code is a lot more centralized and jazz and i can do a lot more stuff

the lines cut down are deceptive btw because i had to chance a lot of datum stuff to datum/component this probably halved the snowflake code shit i had here

also uh descriptions don't output upgrades you've maxed out now which is nice

basically this is for setting the framework for the mob stuff and fixes turret stuff so i don't have to repeatedly go like "ok the turret is shooting" to every single upgrade" basically for the mob stuff "ok the thing moved ok the thing hit the thing" components are great

why am i still typing i must be really bored making gamemodes is hard

Whoneedspacee committed:
moves before fire signal to base of projectiles

removes eh beacon phrases that probably will be made later to be more accurate and helpful

adds component to prevent things from crossing the beacon walls, kills them if they're living and deletes them if they're not living

adds many different spore types and upgrades to them as well as their own upgrades

sentient spores no longer take extra damage from being exposed to high temperatures (because the infection does not work like blob and its stupid)

changes layer of infection to stop covering up things

upgrades now bold their name when being displayed

adds icons for all the new types of spores

Whoneedspacee committed:
spores and such no longer die in lockers in objects or while respawning while past the beacon walls

respawn times no longer get reset when spores change type

Whoneedspacee committed:
changes how infection cores deal damage

fixes up legendary weapons code to not look like complete shit

excaliju no longer does bonus force to all structures

replaces excaliju icon with a way more badass one

renames bosses file to events file since it's going to be more than just bosses now

Whoneedspacee committed:
infection is now actually a gamemode woooooooooo

added that baybee to the config probabilities too

reworked the infection to spawn properly as an antagonist as well

also made it clear by making it a global_var and not a list that there is only supposed to be one infection alive at a time

also made it so that commanders cannot be spawned in the initialize if there's already another commander just to make sure

also sets up the base for the doom clock events for the infection legendary weapon drops and bosses and such

Whoneedspacee committed:
infection now works properly as a gamemode

you no longer die if your loc is an object

adds events for infection

adds time to next event for events subsystem formatted

no longer double adds the key to the new antagonist or deletes the previous mob they were in

infection spores now directly spawn into the commanders contents unless there's an infection core and then it spawns them directly there

adds safety checking for infection core to not spawn multiple times

the infection core on spawning now removes sentient spores from the commanders contents

fixes runtime when rotating through infection nodes while respawning

mobs now have a stat which tells them when the next infection doom event is coming

Whoneedspacee committed:
fixes upgraded sentient spores not getting selected for doom clock events or being in the mobs list

adds debug option to work with gamemode players pick

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into infectionnew

# Conflicts:
#	_maps/map_files/BoxStation/BoxStation.dmm

Whoneedspacee committed:
no more reports interrupting already existing reports in infection gamemode

changes placement times on the core and messages to be less overlapped

changes doom event messages so they don't overlap

adds a pre-sound effect to the doom event

removes the fact that infection cannot expand to space

changes some blob icons

nemvar commented:
first

Shadowflame909 commented:
Hell yeah. I can already see tons of ghost role fun.

But I can also see a bunch of people suiciding for said ghost role fun.

Whoneedspacee commented:
I’m planning on ghost role fun for the station too in terms of centcomm / nanotrasen reinforcements. Not totally sure how I want to implement that just yet, if you have any ideas let me know.

habader commented:
wew lad

optimumtact commented:
![](https://cdn.discordapp.com/attachments/494190398286069770/576955456233275392/infection.jpg)

optimumtact commented on code/game/gamemodes/infection/infection.dm:
```diff
@@ -0,0 +1,69 @@
+	var/spore_datum_type = /datum/antagonist/infection/spore
+
+/datum/game_mode/infection/pre_setup()
+	if(!GLOB.infection_spawns.len)
```
why is this a global list?

Whoneedspacee replied:
It’s the spawn positions which are waypoints in the map. It’s used by the antagonist to spawn in the powers file. Similar to what’s done here https://github.com/tgstation/tgstation/blob/25f18ff5232f6db4ec6806535eed4a98e11e984b/code/modules/antagonists/nukeop/nukeop.dm#L97

Tlaltecuhtli commented:
blob and war ops? just what i needed

MrFluffster commented:
Please use the enter key, it really helps with reading when the text is broken up into manageable chunks.
Nobody likes blocks of text.

knacker48 commented:
If this comes in will it completely replace blob as it is now?

wesoda25 commented:
I don’t really have an opinion on most of this, but I think it’s important to note the waiting period is the reason a lot of people hate nuke ops and had it set to such a low chance. A boring grace period followed by an antag which seems even less flexible than nuke ops doesn’t seem like it’ll age well.

davethwave commented:
Might be good to replace the turret sprites, think something like the anti air turret things from the zerg could be good inspiration to go after with how this gamemode works.

Whoneedspacee commented:
pretty much everything is placeholder  in terms of icons

optimumtact commented:
Please reduce your changelog to the first line, don't try to stuff a gamemode tutorial in it, that's what the wiki is for.

Whoneedspacee commented:
got it

OverlyGeneric commented:
This is like the Purge but for powergamers to go all out just to get out their system.

Whoneedspacee committed:
removes stupid analyzer code

removes ctrl + click shields

removes old hud creation stuff

adds actions for creation and overmind abilities instead for unlockability and just more easy organization and new adding of powers

fixes round ending stuff for the gamemode

fixes roundend report weirdness

the infection no longer starts with the ability to make a turret

renames and makes core turrets more unique rather than just max upgraded resistance turrets

you can no longer place stuff too close to the core (prevents gimmicky stuff like removing core turrets to place your own turrets, etc)

stops using transfer_to in a lot of places because they transfer abilities and powers

checks if minion types exist or boss types exist before trying to create them for doom clock events (boss only, minion only events?)

infection mob upgrade menu now cuts to the case, printing out the upgrades list and opening the menu at the same time instead of having a ton of options

Whoneedspacee commented:
> I don’t really have an opinion on most of this, but I think it’s important to note the waiting period is the reason a lot of people hate nuke ops and had it set to such a low chance. A boring grace period followed by an antag which seems even less flexible than nuke ops doesn’t seem like it’ll age well.

I'm very much considering lowering the waiting period to an extremely low amount such as 2 minutes with the excuse that they were unable to track it until it entered the solar system, and just making the infection much weaker off the bat. It would provide much more of a tug of war and slow build up for both sides. Especially since the defensive beacons along with weaker enemies makes it harder for the infection to advance.

I also plan on making it easier to fight the infection by giving gravity while you're inside it as well as not requiring any atmospheric suits while you're inside it as well, probably something the infection could change with an evolution for some added variability.

By very much considering I mean all future design is going towards this unless I find some major problem with it.

Whoneedspacee commented:
damn you blob refactor fucking my shit up

Whoneedspacee committed:
adds component signals for infection specific stuff

changes point rates and point caps to be less massive by just reducing everything

adds upgrades for resource and factory structures

changes the point returns of every structure

refactors upgrade system so structure upgrades derive from structure type

upgrade descriptions now print regardless of whether they've been fully bought or not

fixes an issue with there being no name or description for homing bullets upgrade

Whoneedspacee committed:
cool entering and exiting effect for infection as well as lowering your vision in the infection making combat more difficult (might consider increasing delay of fade in and fade out to make more punishing)

Whoneedspacee committed:
adds on screen hud blocking vision when you're inside the infection

adds a forced gravity that ignores space in the component

adds a variable to atoms_movable which makes it so that object can be used as a space backup to grab onto even if it is not dense or you cant pass through it (think space vines, can pass through just slowly)

the gamemode no longer checks if beacons are destroyed since the infection commander code will handle that regardless

moves beacon crossing component to actual components folder

infection structures can now have a building time so you can't just instantly box someone in with shields for example

adds actual core types so they can be handled seperately from normal types

adds a max upgrade proc for instant max upgrades

ninjanomnom commented on code/datums/components/nobeaconcrossing.dm:
```diff
@@ -0,0 +1,43 @@
+/datum/component/no_beacon_crossing
+	var/atom/parentatom
```
Why do people keep trying to do this, stop it. Typecast when you need it don't store your parent again

ninjanomnom commented on code/datums/components/nobeaconcrossing.dm:
```diff
@@ -0,0 +1,43 @@
+	var/atom/parentatom
+
+/datum/component/no_beacon_crossing/Initialize()
+	parentatom = parent
```
cease

ninjanomnom commented on code/datums/components/chasm.dm:
```diff
@@ -51,8 +51,13 @@
+	if(locate(/obj/structure/infection) in parent.contents)
+		// no thanks that would be stupid to fall in over that or after destroying it
+		var/turf/T = get_turf(parent)
+		T.ChangeTurf(/turf/open/floor/holofloor/asteroid)
```
Please don't put snowflake code in components like this, why is this here?

Whoneedspacee committed:
no more snowflake checks for chasms in the component, now the infection checks on being pulsed if there's a chasm and changes it to an empty space turf if it is

no more components storing their parent in a different var, also removed a lot of worthless code in the upgrades file

the core has more spread of a defense than before

comsig life for infection is now sent in the base proc instead of being spread out everywhere stupidly

infection mobs now generate their own upgrade points as well as have a hud for them instead of sharing them with the overmind

you can now override the structure build time for infection structures

infection mobs now properly update their icons on life

ninjanomnom commented on code/modules/antagonists/infection/infection/theinfection.dm:
```diff
@@ -0,0 +1,388 @@
+	Life()
+
+/obj/structure/infection/proc/Life()
+	SEND_SIGNAL(src, COMSIG_INFECTION_LIFE_TICK)
```
Absolutely not allowed, no signals in processing procs.

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/upgrades.dm:
```diff
@@ -0,0 +1,398 @@
+/datum/component/infection/upgrade
```
components go in the component folder

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/upgrades.dm:
```diff
@@ -0,0 +1,398 @@
+		register_signals()
+
+/datum/component/infection/upgrade/proc/register_signals()
+	RegisterSignal(parent, COMSIG_INFECTION_LIFE_TICK, .proc/life_tick)
```
If something needs to process it should not piggyback off another thing processing. Add it to a subsystem on its own if it needs to process.

ninjanomnom commented:
Contact me in discord/irc/whatever, there's a lot I need to talk to you about in regards to your use of components here

Whoneedspacee committed:
new epic icons & moves infection icons back to their own file

smoothing is coming for the epic icons

no more core smoke as it doesn't fit the theme of the infection anymore (crystal)

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into infectionnew

# Conflicts:
#	code/__DEFINES/components.dm
#	code/modules/projectiles/projectile.dm

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into infectionnew

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into infectionnew

Whoneedspacee committed:
icon changes

commander no longer calls life

announcement changes

upgrades system moved to straight datum and adds components

Whoneedspacee committed:
wait what are those even supposed to be there

Whoneedspacee committed:
lots of weird diffs just kind of helping my sanity to make sure im not messing something up

ninjanomnom commented on code/game/gamemodes/infection/infection.dm:
```diff
@@ -0,0 +1,71 @@
+	// end if commander is dead, or if beacons are destroyed (handled in antagonist win check)
+	if(C)
+		return FALSE
+	return TRUE
```
`return !isnull(GLOB.infection_commander)`

ninjanomnom commented on code/modules/antagonists/infection/beacon.dm:
```diff
@@ -0,0 +1,158 @@
+	icon_state = "beacon_start"
+
+/obj/effect/landmark/beacon_start/Initialize(mapload)
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/antagonists/infection/beacon.dm:
```diff
@@ -0,0 +1,158 @@
+	var/list/walls = list()
+
+/obj/structure/beacon_generator/Initialize(mapload)
+	START_PROCESSING(SSobj,src)
```
Why do this before the parent call?

ninjanomnom commented on code/modules/antagonists/infection/beacon.dm:
```diff
@@ -0,0 +1,158 @@
+					INF.take_damage(1000, BRUTE, "bomb", 0)
+			sleep(4)
+	else
+		playsound(src.loc, 'sound/effects/empulse.ogg', 300, 1, 10, pressure_affected = FALSE)
```
```
if(obj_integrity > 0)
  playsound(...
  return
otherstuff...
```

ninjanomnom commented on code/modules/antagonists/infection/infection/commander.dm:
```diff
@@ -0,0 +1,238 @@
+
+/mob/camera/commander/Move(NewLoc, Dir = 0)
+	if(meteor)
+		return 0
```
Use FALSE and TRUE always for bools
```suggestion
		return FALSE
```

ninjanomnom commented on code/modules/antagonists/infection/infection/commander.dm:
```diff
@@ -0,0 +1,238 @@
+	if(meteor)
+		return 0
+	forceMove(NewLoc)
+	return 1
```
```suggestion
	return TRUE
```

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/core.dm:
```diff
@@ -0,0 +1,168 @@
+	SSevents.frequency_upper = DOOM_CLOCK_EVENT_DELAY
+	SSevents.toggleInfectionmode()
+	SSevents.reschedule()
+	. = ..()
```
Please have the parent call as early as possible in the proc

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/core.dm:
```diff
@@ -0,0 +1,168 @@
+	add_overlay(mutable_appearance('icons/mob/blob.dmi', "blob_core_overlay"))
+
+/obj/structure/infection/core/Destroy()
+	. = ..()
```
...except destroy which needs to be last

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/core.dm:
```diff
@@ -0,0 +1,168 @@
+	for(var/i = 1 to 9)
+		for(var/atom/A in urange(i, explodeloc) - urange(i - 1, explodeloc))
+			var/power = CEILING(i/3, 1)
+			A.ex_act(power)
```
Why are you handrolling your own explosion?

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/core.dm:
```diff
@@ -0,0 +1,168 @@
+			if(istype(A, /obj/structure/infection))
+				var/obj/structure/infection/INF = A
+				INF.take_damage(600 / power, BRUTE, "bomb", 0)
+		sleep(4)
```
Switch this to be multiple timers

ninjanomnom reviewed (changes requested):
Still more but I've been interrupted

ninjanomnom commented on code/game/gamemodes/infection/infection.dm:
```diff
@@ -0,0 +1,71 @@
+		return FALSE
+	var/n_spores = min(CEILING(num_players() / 10, 1), antag_candidates.len)
+	if(n_spores >= required_enemies || GLOB.Debug2)
+		for(var/i = 0, i < n_spores, ++i)
```
Or just `1 to n_spores` if you change the other stuff

FuryMcFlurry commented:
![image](https://user-images.githubusercontent.com/8770071/58988177-54b43b80-87e1-11e9-924e-d87cb3a3a8f0.png)

Blobs are dum, crystals are cool.

Whoneedspacee commented:
thank fury for the new icons he's doing

Whoneedspacee committed:
1 is true and 0 is false

forced_gravity component subtypes are stupid

proper loops and checking formatting in the gamemode

initialize as early as possible and destroy as late as possible

default infection color is now pure white

core now has a big 96x96 icon

no more core snowflake explosion handling

upgrade_types mostly replaced with upgrade_subtype which is much neater and less work when new upgrades are made

icon directories and files moved around again

point return for resource generators is now functional

shields now smooth properly with themselves and have much more dull colors for the crystals as well as adding a missed crystal type

Whoneedspacee committed:
typesof god im an idiot

zxaber commented:
>The shadow realm has come to claim the station

Looks good, for real

ninjanomnom commented on code/datums/components/nobeaconcrossing.dm:
```diff
@@ -0,0 +1,40 @@
+/datum/component/no_beacon_crossing/Initialize()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/check_passed)
```
`return COMPONENT_INCOMPATIBLE` when parent isnt a movable

ninjanomnom commented on code/datums/components/nobeaconcrossing.dm:
```diff
@@ -0,0 +1,40 @@
+		should_die = TRUE
+	if(isobj(parentatom.loc))
+		should_die = FALSE // don't kill them if they're in a locker, or something is holding them
+	if(should_die)
```
Make this a seperate proc like kill_parent() or something

ninjanomnom replied:
You've made the proc, now you dont need to set should_die and can just directly call `kill_parent()`

ninjanomnom commented on code/modules/antagonists/infection/beacon.dm:
```diff
@@ -0,0 +1,158 @@
+	qdel(src)
+	for(var/i = 1 to 5)
+		for(var/atom/A in urange(i, explodeloc) - urange(i - 1, explodeloc))
+			A.ex_act(EXPLODE_LIGHT)
```
Another case of "why are you handrolling an explosion"

ninjanomnom commented on code/modules/antagonists/infection/beacon.dm:
```diff
@@ -0,0 +1,158 @@
+			if(istype(A, /obj/structure/infection))
+				var/obj/structure/infection/INF = A
+				INF.take_damage(1000, BRUTE, "bomb", 0)
+		sleep(4)
```
5 timers 4 ticks apart each instead of this sleep

ninjanomnom commented on code/modules/antagonists/infection/infection.dm:
```diff
@@ -0,0 +1,64 @@
+	var/point_rate_human_infection = 2
+
+/datum/antagonist/infection/roundend_report()
+	var/basic_report = ..()
```
You're allowed to change `.` you know. `. = ..()` works just fine here

ninjanomnom commented on code/modules/antagonists/infection/infection.dm:
```diff
@@ -0,0 +1,64 @@
+	var/turf/start = pick(GLOB.infection_spawns)
+	var/mob/newmob = create_mob_type(start)
+	owner.transfer_to(newmob, TRUE)
+	. = ..()
```
If this needs to be the last line do `return ..()` instead

ninjanomnom commented on code/modules/antagonists/infection/infection/commander.dm:
```diff
@@ -0,0 +1,238 @@
+	for(var/type_action in default_actions)
+		var/datum/action/innate/infection/add_action = new type_action()
+		add_action.Grant(src)
+	.= ..()
```
Why is this so late in initialize? None of this looks like it needs to be before parent inits?

ninjanomnom commented on code/modules/antagonists/infection/infection/commander.dm:
```diff
@@ -0,0 +1,238 @@
+	return ..()
+
+/mob/camera/commander/Login()
+	..()
```
default to `. = ..()`
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/antagonists/infection/infection/commander.dm:
```diff
@@ -0,0 +1,238 @@
+	add_points(0)
+
+/mob/camera/commander/examine(mob/user)
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/antagonists/infection/infection/infections/core.dm:
```diff
@@ -0,0 +1,165 @@
+	update_icon() //so it atleast appears
+	if(!placed && !overmind)
+		return INITIALIZE_HINT_QDEL
+	. = ..()
```
Again, why is this so late in initialize?

Whoneedspacee committed:
checks movableatom for beacon crossing component

doom event mobs now die once a new event is created

no more snowflake explosions

adds more actions and makes them have cooldown timers

the infection commander is now added to the z level client mobs list and as a result can now hear sounds properly

the infection now expands more cardinally

infections that are building now don't turn into blank space while they are building

ArcaneMusic commented:
Holy shit the sprites look really good, like damn. Some of the textures for walls and floors could maybe look a bit more jagged, perhaps, but I'm still really pleased with the theming for this gamemode being realized.

Whoneedspacee committed:
commander processing is no longer literally broken

upgraded infection mob types now retain their spent points

fixes icon for unlock turret ability

Whoneedspacee commented:
https://www.youtube.com/watch?v=s1a1o8RxLpw&feature=youtu.be

Another update video, been a while eh?

Whoneedspacee committed:
lots of filename and directory changes mimicking the recent blob refactor in order to make working with the files less of a pain

adds new icons for infection node that are animated when they actually pulse

fixes issue with radial menus not appearing on the correct position

building infections now take the pixel positions of their icon so they're on the proper position

infection turrets no longer have the wrong blob overlay

SpaceManiac commented:
Closing as Stale for being Work In Progress for over a week with no updates. If you want to continue work, reply here or ping a maintainer.
